rustic-cargo.el

```diff
@@ -458,13 +466,69 @@ If running with prefix command `C-u', read whole command from minibuffer."
                              'compile-history)))))
     (rustic-run-cargo-command command (list :mode 'rustic-cargo-run-mode))))
 
-(define-derived-mode rustic-cargo-run-mode rustic-compilation-mode "Cargo run"
-  "Mode for 'cargo run' that derives from `rustic-compilation-mode', but uses
-the keymap of `comint-mode' so user input is possible."
+(defun rustic-cargo-run-mode ()
+  (interactive)
+  (if rustic-cargo-run-use-comint
+      ;; rustic-cargo-comint-run-mode toggles the mode; we want to
+      ;; always enable.
+      (unless (and (boundp 'polymode-mode)
+                   polymode-mode
+                   (memq major-mode '(rustic-cargo-plain-run-mode
+                                      comint-mode)))
+        (rustic-cargo-comint-run-mode))
+    (rustic-cargo-plain-run-mode)))
+
+(define-derived-mode rustic-cargo-plain-run-mode rustic-compilation-mode "Cargo run"
+  "Mode for 'cargo run' that derives from `rustic-compilation-mode'.
+
+To send input to the compiled program, use
+`rustic-compile-send-input'.  If you set
+`rustic-cargo-run-use-comint' to t, you can also just type in a
+string and hit RET to send it to the program.  The latter
+approach requires installing polymode."
   (buffer-disable-undo)
   (setq buffer-read-only nil)
   (use-local-map comint-mode-map))
 
+(defun rustic-cargo-comint-run-mode ()
+  "Mode for 'cargo run' that combines `rustic-compilation-mode' with `comint-mode',
+the former for highlighting and interacting with compiler errors,
+and the latter for interacting with the compiled program."
+  ;; First time around, define the mode and invoke it.  Next time, the
+  ;; symbol will have been overwritten so this runs only once.
+  (require 'polymode)
```

Comment:
Maybe we should check if polymode can be required. If not, we can inform the user about the missing package with a message instead of an elisp error.

Reply:
Good point.  Fix added.